source/game/system/KPadController.cc

```diff
@@ -236,6 +301,87 @@ u8 KPadGhostTrickButtonsStream::readVal() const {
     return currentSequence >> 0x8 & ~0x80;
 }
 
+/* ================================ *
+ *     HOST CONTROLLER
+ * ================================ */
+
+KPadHostController::KPadHostController() = default;
+
+KPadHostController::~KPadHostController() = default;
+
+ControlSource KPadHostController::controlSource() {
+    return ControlSource::Host;
+}
+
+void KPadHostController::reset(bool driftIsAuto) {
+    m_driftIsAuto = driftIsAuto;
+    m_raceInputState.reset();
+    m_connected = true;
+}
+
+/// @brief Sets the inputs of the controller.
+/// @param state The specified inputs packaged in the state. Only buttons, stick, and trick matter.
+/// @return Input state validity.
+bool KPadHostController::setInputs(const RaceInputState &state) {
+    return setInputs(state.buttons, state.stick, state.trick);
+}
+
+/// @brief Sets the inputs of the controller.
+/// @param buttons The button inputs.
+/// @param stick The stick inputs, as a 2D vector.
+/// @param trick The trick input.
+/// @return Input state validity.
+bool KPadHostController::setInputs(u16 buttons, const EGG::Vector2f &stick, Trick trick) {
+    m_raceInputState.buttons = buttons;
+    m_raceInputState.stick = stick;
+    m_raceInputState.trick = trick;
+
+    return m_raceInputState.isValid();
+}
+
+/// @brief Sets the inputs of the controller.
+/// @param buttons The button inputs.
+/// @param stickX The stick input on the X axis.
+/// @param stickY The stick input on the Y axis.
+/// @param trick The trick input.
+/// @return Input state validity.
+bool KPadHostController::setInputs(u16 buttons, f32 stickX, f32 stickY, Trick trick) {
+    m_raceInputState.buttons = buttons;
+    m_raceInputState.stick.x = stickX;
+    m_raceInputState.stick.y = stickY;
+    m_raceInputState.trick = trick;
+
+    return m_raceInputState.isValid();
+}
+
+/// @brief Sets the inputs of the controller.
+/// @details A different name is specified to avoid any ambiguity with the parameters.
+/// @param buttons The button inputs.
+/// @param stickXRaw The 7-centered raw stick input on the X axis.
+/// @param stickYRaw The 7-centered raw stick input on the Y axis.
+/// @param trick The trick input.
+/// @return Input state validity.
+bool KPadHostController::setInputsRawStick(u16 buttons, u8 stickXRaw, u8 stickYRaw, Trick trick) {
+    return setInputs(buttons, (static_cast<f32>(stickXRaw) - 7.0f) / 7.0f,
+            (static_cast<f32>(stickYRaw) - 7.0f) / 7.0f, trick);
```

Comment:
There are several instances where we have `(static_cast<f32>(val) - 7.0f) / 7.0f` in order to normalize the input. Could we maybe make this a constexpr helper function or something?

Reply:
Added.